File: Blowfish/java/src/com/nokinobi/Decryptor.java
package com.nokinobi;

import com.nokinobi.stream.LittleEndianInputStream;
import com.nokinobi.stream.LittleEndianOutputStream;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import com.nokinobi.functions.Functions;
import java.util.Arrays;
import java.util.Scanner;


public class Decryptor {


    public static void Decrypt(String encryptedFile,String decryptedFile) throws IOException {

        LittleEndianInputStream reader = new LittleEndianInputStream(new FileInputStream(encryptedFile));
        LittleEndianOutputStream writer = new LittleEndianOutputStream(new FileOutputStream(decryptedFile));

        int[] keys = Functions.ReadKeys(reader);
        int[][] matrix = Functions.ReadMatrix(reader);
        int fileLen = reader.readInt();


        System.out.println("Count of rounds = ");
        Scanner scanner = new Scanner(System.in);
        int roundsCount = scanner.nextInt();

        keys = Arrays.copyOf(keys,roundsCount);
        Reverse(keys);

        Tuple block = new Tuple();
        for(int i=0;i< fileLen/8;++i){
            Functions.ReadBlock(block, reader);
            Functions.ProcessBlock(block, Arrays.copyOf(keys,roundsCount), matrix);
            writer.write(block.le);
            writer.write(block.re);
        }

        int tail = (fileLen + 8) % 8;
        if(tail != 0){
            Functions.ReadBlock(block, reader);
            Functions.ProcessBlock(block, Arrays.copyOf(keys,roundsCount), matrix);
            long pi = ( ((long)block.re)<<32 ) + block.le;
            for(int i=0;i<tail;++i)
              writer.write( (byte)(pi>>>i*8 ) );
        }

        reader.close();
        writer.close();
    }

    public static void Reverse(int[] arr){
        for(int i = 0; i < arr.length / 2; i++)
        {
            int temp = arr[i];
            arr[i] = arr[arr.length - i - 1];
            arr[arr.length - i - 1] = temp;
        }
    }
}

File: Blowfish/c#/Blowfish/Decryptor.cs
using System;
using System.IO;
using System.Net;
using functions;
using Tuple = functions.Tuple;
using static functions.Functions;

namespace Blowfish
{
    class Decryptor
    {
        public static void Decrypt(string encryptedFile, string decryptedFile)
        {
            FileStream inputStream = new FileStream(encryptedFile,FileMode.Open);
            BinaryReader reader= new BinaryReader(inputStream);
            BinaryWriter writer= new BinaryWriter(new FileStream(decryptedFile, FileMode.OpenOrCreate));

            int[] keys = ReadKeys(reader);
            int[,] matrix = ReadMatrix(reader);
            
             
            Console.Write("Count of rounds = ");
            int roundsCount = Convert.ToInt32(Console.ReadLine());
            Array.Resize(ref keys,roundsCount);
            Array.Reverse(keys);
            
            int fileLength = reader.ReadInt32();
            
            Tuple tuple = new Tuple();
            for (int i = 0; i < fileLength / 8; i++)
            {
                ReadBlock(tuple, inputStream);
                ProcessBlock(tuple, keys, matrix,roundsCount);
                writer.Write(tuple.le);
                writer.Write(tuple.re);
            }

            int tail = fileLength % 8;
            if (tail > 0)
            {
                ReadBlock(tuple, inputStream);
                ProcessBlock(tuple, keys, matrix,roundsCount);
                long pi = ((long) tuple.re << 32) + tuple.le;
                for (int i = 0; i < tail; ++i)
                    writer.Write(new[] {(byte) (pi >> i * 8)}, 0, 1);
            }
            inputStream.Close();
            writer.Close();
        }
    }
}